

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>src.parsing &mdash; Fold U 2.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> Fold U
          

          
          </a>

          
            
            
              <div class="version">
                2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../alignment.html">Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../template.html">Template</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../query.html">Query</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../residue.html">Residue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../atom.html">Atom</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../score.html">Score</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../parsing.html">Parsing</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Fold U</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>src.parsing</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for src.parsing</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">.. module:: parsing</span>
<span class="sd">   :synopsis: This module implements all the functions to parse either input</span>
<span class="sd">                or additional necessary files.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># Third-party modules</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="k">import</span> <span class="n">seq1</span>
<span class="kn">from</span> <span class="nn">conkit.applications</span> <span class="k">import</span> <span class="n">CCMpredCommandline</span>
<span class="c1"># Local modules</span>
<span class="kn">from</span> <span class="nn">src.residue</span> <span class="k">import</span> <span class="n">Residue</span>
<span class="kn">from</span> <span class="nn">src.alignment</span> <span class="k">import</span> <span class="n">Alignment</span>
<span class="kn">from</span> <span class="nn">src.query</span> <span class="k">import</span> <span class="n">Query</span>
<span class="kn">from</span> <span class="nn">src.template</span> <span class="k">import</span> <span class="n">Template</span>


<div class="viewcode-block" id="parse_metafold"><a class="viewcode-back" href="../../parsing.html#src.parsing.parse_metafold">[docs]</a><span class="k">def</span> <span class="nf">parse_metafold</span><span class="p">(</span><span class="n">metafold_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts name of metafold as key and associated pdb as</span>
<span class="sd">        value in dictionary.</span>

<span class="sd">        Args:</span>
<span class="sd">            metafold_file (str): A METAFOLD.list file containing for a given line a</span>
<span class="sd">                           template name and a pdb file name associated.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary with key = template name and value = pdb file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">metafold_dict</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># Initalization of the dictionary</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">metafold_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">metafold_dict</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">metafold_dict</span></div>


<div class="viewcode-block" id="parse_foldrec"><a class="viewcode-back" href="../../parsing.html#src.parsing.parse_foldrec">[docs]</a><span class="k">def</span> <span class="nf">parse_foldrec</span><span class="p">(</span><span class="n">foldrec_file</span><span class="p">,</span> <span class="n">nb_templates</span><span class="p">,</span> <span class="n">metafold_dict</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts the score, the template name, the query and template sequences for</span>
<span class="sd">        the first n alignments from a file containing N profil-profil alignments and</span>
<span class="sd">        creates a list of Alignment objects. It also gives the template&#39;s pdb file name</span>
<span class="sd">        and gets all the coordinates of the CA atoms in the template&#39;s Residue list.</span>

<span class="sd">        Args:</span>
<span class="sd">            foldrec_file (str): The file containing N profil-profil alignments and their</span>
<span class="sd">                                corresponding scores.</span>
<span class="sd">            nb_templates (int): Number of alignments to retrieve from the file and chosen</span>
<span class="sd">                                by the user.</span>
<span class="sd">            metafold_dict (dict): A dictionary with key = template name</span>
<span class="sd">                                        and value = pdb file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary with key = template name and value = an Alignment object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Regex :</span>
    <span class="n">num_reg</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;^No</span><span class="se">\\</span><span class="s2">s*([0-9]+)&quot;</span><span class="p">)</span>
    <span class="n">template_name_reg</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;Alignment :.*vs</span><span class="se">\\</span><span class="s2">s+([A-Za-z0-9-_]+)&quot;</span><span class="p">)</span>
    <span class="n">score_reg</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;^Score :</span><span class="se">\\</span><span class="s2">s+([-0-9</span><span class="se">\\</span><span class="s2">.]+)&quot;</span><span class="p">)</span>
    <span class="n">query_seq_reg</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;^Query</span><span class="se">\\</span><span class="s2">s*([0-9]+)</span><span class="se">\\</span><span class="s2">s*([A-Z0-9-]+)</span><span class="se">\\</span><span class="s2">s*([0-9]+)&quot;</span><span class="p">)</span>
    <span class="n">template_seq_reg</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;^Template</span><span class="se">\\</span><span class="s2">s*[0-9]+</span><span class="se">\\</span><span class="s2">s*([A-Z-]+)&quot;</span><span class="p">)</span>
    <span class="n">empty_query_reg</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;^Query</span><span class="se">\\</span><span class="s2">s+</span><span class="se">\\</span><span class="s2">d</span><span class="se">\\</span><span class="s2">s-+</span><span class="se">\\</span><span class="s2">s+</span><span class="se">\\</span><span class="s2">d.*$&quot;</span><span class="p">)</span>

    <span class="n">alignment_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">count_templates</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">foldrec_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">query_reg_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">template_reg_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
            <span class="c1"># The loop is break when the required nb of templates is reached :</span>
            <span class="k">if</span> <span class="n">count_templates</span> <span class="o">==</span> <span class="n">nb_templates</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="c1"># Search a regex for the current line :</span>
            <span class="n">num_found</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">num_reg</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="n">template_name_found</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">template_name_reg</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="n">score_found</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">score_reg</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="n">query_seq_found</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">query_seq_reg</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="n">template_seq_found</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">template_seq_reg</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="n">empty_query_found</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">empty_query_reg</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="c1"># A num is found :</span>
            <span class="k">if</span> <span class="n">num_found</span><span class="p">:</span>
                <span class="n">num</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">num_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
            <span class="c1"># A template name is found :</span>
            <span class="k">if</span> <span class="n">template_name_found</span><span class="p">:</span>
                <span class="c1"># These templates have more than 1 chain, so we must skip them</span>
                <span class="k">if</span> <span class="n">template_name_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;bromodomain&quot;</span><span class="p">,</span> <span class="s2">&quot;rhv&quot;</span><span class="p">,</span> <span class="s2">&quot;Peptidase_A6&quot;</span><span class="p">,</span> <span class="s2">&quot;ins&quot;</span><span class="p">,</span>
                                                    <span class="s2">&quot;Arg_repressor_C&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_decarbox&quot;</span><span class="p">,</span> <span class="s2">&quot;prc&quot;</span><span class="p">,</span>
                                                    <span class="s2">&quot;Chorismate_mut&quot;</span><span class="p">]:</span>
                    <span class="c1"># We skip the alignment</span>
                    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
                        <span class="nb">next</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="n">template_name</span> <span class="o">=</span> <span class="n">template_name_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="c1"># A score is found :</span>
            <span class="k">if</span> <span class="n">score_found</span><span class="p">:</span>
                <span class="n">score</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">score_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
            <span class="c1"># Empty alignement (the query = gaps only)</span>
            <span class="k">if</span> <span class="n">empty_query_found</span> <span class="ow">and</span> <span class="n">query_reg_count</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Skipping alignement &quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;. The query is only composed of gaps.&quot;</span><span class="p">)</span>
                <span class="c1"># We skip the alignment</span>
                <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span>
                    <span class="nb">next</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="c1"># A query sequence is found :</span>
            <span class="k">if</span> <span class="n">query_seq_found</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">query_reg_count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">query_first</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">query_seq_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
                    <span class="n">query_seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">Residue</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">query_seq_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">))]</span>
                    <span class="n">query_last</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">query_seq_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">))</span>
                    <span class="n">query_reg_count</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">query_reg_count</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">sec_struct</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">query_seq_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">))):</span>
                        <span class="n">query_seq</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span><span class="o">.</span><span class="n">secondary_struct</span> <span class="o">=</span> <span class="n">sec_struct</span>
                    <span class="n">query_reg_count</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">query_reg_count</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">ss_conf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">query_seq_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">))):</span>
                        <span class="k">if</span> <span class="n">ss_conf</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                            <span class="n">ss_conf</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ss_conf</span><span class="p">)</span>
                        <span class="n">query_seq</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span><span class="o">.</span><span class="n">ss_confidence</span> <span class="o">=</span> <span class="n">ss_conf</span>
                    <span class="n">query_reg_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># A template sequence is founds :</span>
            <span class="k">if</span> <span class="n">template_seq_found</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">template_reg_count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">template_seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">Residue</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">template_seq_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))]</span>
                    <span class="n">template_reg_count</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">template_reg_count</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">sec_struct</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">template_seq_found</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))):</span>
                        <span class="n">template_seq</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span><span class="o">.</span><span class="n">secondary_struct</span> <span class="o">=</span> <span class="n">sec_struct</span>
                    <span class="n">template_reg_count</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="c1"># Add a new alignment object in the list :</span>
                    <span class="n">ali</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">num</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span>
                                    <span class="n">Query</span><span class="p">(</span><span class="n">query_seq</span><span class="p">,</span> <span class="n">query_first</span><span class="p">,</span> <span class="n">query_last</span><span class="p">),</span>
                                    <span class="n">Template</span><span class="p">(</span><span class="n">template_name</span><span class="p">,</span> <span class="n">template_seq</span><span class="p">))</span>
                    <span class="n">ali</span><span class="o">.</span><span class="n">template</span><span class="o">.</span><span class="n">set_pdb_name</span><span class="p">(</span><span class="n">metafold_dict</span><span class="p">)</span>
                    <span class="n">ali</span><span class="o">.</span><span class="n">template</span><span class="o">.</span><span class="n">parse_pdb</span><span class="p">(</span><span class="s2">&quot;data/pdb/&quot;</span><span class="o">+</span><span class="n">ali</span><span class="o">.</span><span class="n">template</span><span class="o">.</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot;/&quot;</span><span class="o">+</span><span class="n">ali</span><span class="o">.</span><span class="n">template</span><span class="o">.</span><span class="n">pdb</span>
                                           <span class="o">+</span><span class="s2">&quot;.atm&quot;</span><span class="p">)</span>
                    <span class="n">alignment_dict</span><span class="p">[</span><span class="n">template_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">ali</span>
                    <span class="n">count_templates</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">alignment_dict</span></div>


<div class="viewcode-block" id="parse_benchmark"><a class="viewcode-back" href="../../parsing.html#src.parsing.parse_benchmark">[docs]</a><span class="k">def</span> <span class="nf">parse_benchmark</span><span class="p">(</span><span class="n">query_name</span><span class="p">,</span> <span class="n">benchmark_file</span><span class="p">,</span> <span class="n">alignment_dict</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extract the line of the benchmark file containing the benchmarks of</span>
<span class="sd">        the studied query. Each benchmark template have an assigned structure</span>
<span class="sd">        corresponding to the degree of similarity with the query</span>
<span class="sd">        (&quot;Family&quot;, &quot;Superfamily&quot; or &quot;Fold&quot;). For a given benchmark template</span>
<span class="sd">        object, the structure is stored in the benchmark item.</span>

<span class="sd">        Args:</span>
<span class="sd">            query_name (str): Name of the query.</span>
<span class="sd">            benchmark_file (str): The path to the benchmark file.</span>
<span class="sd">            alignment_dict (dict): A dictionary with key = template name.</span>
<span class="sd">                                   and value = an Alignment object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">benchmark_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
            <span class="c1"># Only the line containing the query name is extracted</span>
            <span class="k">if</span> <span class="n">query_name</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
                <span class="c1"># List containing for each benchmark the template name and</span>
                <span class="c1"># the associated fold_type</span>
                <span class="n">template_fold_type</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;: &quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">template_fold_type</span><span class="p">:</span>
                    <span class="c1"># The fold_type of the benchmark template is stored in</span>
                    <span class="c1"># the benchmark item</span>
                    <span class="n">template</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">fold_type</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">template</span> <span class="ow">in</span> <span class="n">alignment_dict</span><span class="p">:</span>
                        <span class="n">alignment_dict</span><span class="p">[</span><span class="n">template</span><span class="p">]</span><span class="o">.</span><span class="n">template</span><span class="o">.</span><span class="n">set_benchmark</span><span class="p">(</span><span class="n">fold_type</span><span class="p">)</span>
                <span class="c1"># No need to continue because all benchmarks are stored</span>
                <span class="k">break</span></div>


<div class="viewcode-block" id="parse_dope"><a class="viewcode-back" href="../../parsing.html#src.parsing.parse_dope">[docs]</a><span class="k">def</span> <span class="nf">parse_dope</span><span class="p">(</span><span class="n">dope_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts 30 dope energy values for the 20*20 residus-CA pairs and stores it</span>
<span class="sd">        in a dictionary with key = res_1-res_2 and value = an array of the 30 dope scores.</span>

<span class="sd">        Args:</span>
<span class="sd">            dope_file (str): The file dope.par containing energy values for each</span>
<span class="sd">                       amino acid pair.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary with key = res_1-res_2 and value = an array of</span>
<span class="sd">            the 30 dope energy values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dope_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">dope_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
            <span class="c1"># get the line with C-alpha for both amino acids</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">6</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span> <span class="ow">and</span> <span class="n">line</span><span class="p">[</span><span class="mi">11</span><span class="p">:</span><span class="mi">13</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span><span class="p">:</span>
                <span class="n">res_1</span> <span class="o">=</span> <span class="n">seq1</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">res_2</span> <span class="o">=</span> <span class="n">seq1</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">7</span><span class="p">:</span><span class="mi">10</span><span class="p">])</span>
                <span class="n">dope_dict</span><span class="p">[</span><span class="n">res_1</span><span class="o">+</span><span class="n">res_2</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fromstring</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">14</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">dope_dict</span></div>


<div class="viewcode-block" id="convert_aln_file"><a class="viewcode-back" href="../../parsing.html#src.parsing.convert_aln_file">[docs]</a><span class="k">def</span> <span class="nf">convert_aln_file</span><span class="p">(</span><span class="n">aln_file</span><span class="p">,</span> <span class="n">aln_file_clustal</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Convert a multiple alignment file from fasta to clustal.</span>
<span class="sd">        The first sequence of this multiple alignment is kept and</span>
<span class="sd">        a list of non-gaps position index of this sequence is then generated.</span>

<span class="sd">        Args:</span>
<span class="sd">            aln_file (str): Multiple alignement file (fasta format).</span>
<span class="sd">            aln_file_clustal (str): Multiple alignemnt file (clustal format) to generated.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of indexes corresponding to amino acids positions in aln file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;results/ccmpred/aln/&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Convertion from fasta to clustal</span>
    <span class="n">convert</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span>
        <span class="p">[</span><span class="s2">&quot;./bin/CCMpred/scripts/convert_alignment.py&quot;</span><span class="p">,</span> <span class="n">aln_file</span><span class="p">,</span>
         <span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="n">aln_file_clustal</span><span class="p">],</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># First line of this multiple alignment</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">aln_file_clustal</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">query_seq</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">readline</span><span class="p">()[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="c1"># &quot;non_gap&quot; positions index are saved into a list</span>
        <span class="n">index_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ele</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">query_seq</span><span class="p">)</span> <span class="k">if</span> <span class="n">ele</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">index_list</span></div>


<div class="viewcode-block" id="predict_top_contacts"><a class="viewcode-back" href="../../parsing.html#src.parsing.predict_top_contacts">[docs]</a><span class="k">def</span> <span class="nf">predict_top_contacts</span><span class="p">(</span><span class="n">aln_file</span><span class="p">,</span> <span class="n">index_list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extract N tops couplings based on co_evolution score. co_evolution score</span>
<span class="sd">        is calculated between two non-consecutive amino acids by CCMPRED tool based on</span>
<span class="sd">        MSA multiple alignment. Co_evolution score measures co-occurence of a pair of amino</span>
<span class="sd">        acids in ortholog sequences. Two amino acids have co-evoluated if the</span>
<span class="sd">        occurence of one of them never occur whithout the other.</span>

<span class="sd">        .. math::</span>

<span class="sd">          Nb\_top\_contacts = {length(query) \over 2}</span>

<span class="sd">        Args:</span>
<span class="sd">            aln_file (str): Multiple alignement file (clustal format).</span>
<span class="sd">            index_list (list): List of indexes corresponding to amino acids positions in aln file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary with key = ranking of coupling based on ss_confidence</span>
<span class="sd">            and value = index aa1, index aa2</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Number of top coupling expected</span>
    <span class="n">ntops</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">index_list</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">contact_output</span> <span class="o">=</span> <span class="s2">&quot;data/ccmpred/contact.mat&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s2">&quot;data/ccmpred/&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Run ccmpred : Prediction of contacts</span>
    <span class="n">ccmpred_cline</span> <span class="o">=</span> <span class="n">CCMpredCommandline</span><span class="p">(</span>
        <span class="n">cmd</span><span class="o">=</span><span class="s1">&#39;./bin/CCMpred/bin/ccmpred&#39;</span><span class="p">,</span> <span class="n">alnfile</span><span class="o">=</span><span class="n">aln_file</span><span class="p">,</span> <span class="n">matfile</span><span class="o">=</span><span class="n">contact_output</span><span class="p">)</span>
    <span class="n">ccmpred_cline</span><span class="p">()</span>

    <span class="c1"># Extract tops coupling</span>
    <span class="n">top_couplings</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span>
        <span class="p">[</span><span class="s2">&quot;./bin/CCMpred/scripts/top_couplings.py -n </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span>
         <span class="nb">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">ntops</span><span class="p">),</span> <span class="n">contact_output</span><span class="p">)],</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Create a dictionary of top couplings</span>
    <span class="n">top_couplings_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">top_couplings</span><span class="p">):</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">value</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">index_i</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">index_j</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1"># Do not parse gaps associated with top couplings</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">index_i</span> <span class="ow">in</span> <span class="n">index_list</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">index_j</span> <span class="ow">in</span> <span class="n">index_list</span><span class="p">):</span>
            <span class="n">top_couplings_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">index_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">index_i</span><span class="p">),</span> <span class="n">index_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">index_j</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">top_couplings_dict</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Gabriel Cretin, Hélène Kabbech, Tom Gutman, Flora Mikaeloff, Franz Arnold.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'2.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  <script type="text/javascript" src="../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>